csrc/exceptions.h

```diff
@@ -0,0 +1,268 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <array>
+#include <deque>
+#include <fstream>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <typeinfo>
+#include <vector>
+
+namespace nvfuser {
+
+std::string _get_backtrace(
+    size_t frames_to_skip = 0,
+    size_t maximum_number_of_frames = 64,
+    bool skip_python_frames = true);
+
+struct SourceLocation {
+  const char* function;
+  const char* file;
+  uint32_t line;
+};
+
+std::ostream& operator<<(std::ostream& out, const SourceLocation& loc);
+
+struct CompileTimeEmptyString {
+  operator const std::string&() const {
+    static const std::string empty_string_literal;
+    return empty_string_literal;
+  }
+  operator const char*() const {
+    return "";
```

Comment:
This is being used when the argument list in the macros (user message) is empty to avoid constructing any string or stringstream. If you look at the `_str_wrapper` definitions, they tackle 4 cases -- list of arguments of string type, `std::string`, `char*` and empty argument list.

Reply:
Would you similarly need a static const to return since you're returning a `char*`?

Reply:
^^^ I suspected the same thing. But this one is named as a compile time thing so I'm guessing there's some cpp magic that the life time of a local variable isn't a concern.

Reply:
https://stackoverflow.com/questions/2579874/lifetime-of-a-string-literal-returned-by-a-function

Ha, string literal lives forever apparently. I guess I failed the interview question. :laughing:

Reply:
This is being called within `_str_wrapper` with a `static` return:
```
template <>
struct _str_wrapper<> final {
  static CompileTimeEmptyString call() {
    return CompileTimeEmptyString();
  }
};
```